internal/staging/primitives/pow.go

```diff
@@ -0,0 +1,141 @@
+	// exponent value is decreased by 3 and then multiplied by 8, that
+	// necessarily means that any encoded exponent of 35 or greater will cause
+	// an overflow since 256/8 + 3 = 35.  Decreasing the encoded exponent by one
+	// to 34 results in making 8 bits available, meaning the max mantissa can be
+	// 0xff in that case.  Similarly, decreasing the encoded exponent again to
+	// 33 results in making a total of 16 bits available, meaning the max
+	// mantissa can be 0xffff in that case.  Finally, decreasing the encoded
+	// exponent again to 32 results in making a total of 24 bits available and
+	// since the mantissa only encodes 23 bits, overflow is impossible for all
+	// encoded exponents of 32 or lower.
+	overflows = exponent >= 35 || (exponent >= 34 && mantissa > 0xff) ||
+		(exponent >= 33 && mantissa > 0xffff)
+	if overflows {
+		return n, isSignBitSet, true
+	}
+	n.SetUint64(uint64(mantissa))
+	n.Lsh(8 * (exponent - 3))
+	return n, isSignBitSet, false
+}
+
+// uint256ToDiffBits converts a uint256 to a compact representation using an
+// unsigned 32-bit integer.  The compact representation only provides 23 bits of
+// precision, so values larger than (2^23 - 1) only encode the most significant
+// digits of the number.  See DiffBitsToUint256 for details.
+//
+// NOTE: The only difference between this function and the exported variant is
+// that this one accepts a parameter to indicate whether or not the encoding
+// should be for a negative value whereas the exported variant is always
+// positive as expected for an unsigned value.  This is done in order to stay
+// consistent with the encoding used by legacy code and for testing purposes,
+// however, difficulty bits are only used in Decred to encode unsigned 256-bit
+// integers which represent difficulty targets, so, it will always be called
+// with false in practice.
+func uint256ToDiffBits(n *uint256.Uint256, isNegative bool) uint32 {
+	// No need to do any work if it's zero.
+	if n.IsZero() {
+		return 0
+	}
+
+	// Since the base for the exponent is 256, the exponent can be treated as
+	// the number of bytes it takes to represent the value.  So, shift the
+	// number right or left accordingly.  This is equivalent to:
+	// mantissa /= 256^(exponent-3)
+	var mantissa uint32
+	exponent := uint32((n.BitLen() + 7) / 8)
+	if exponent <= 3 {
+		mantissa = n.Uint32() << (8 * (3 - exponent))
+	} else {
+		// Use a copy to avoid modifying the caller's original value.
+		mantissa = new(uint256.Uint256).RshVal(n, 8*(exponent-3)).Uint32()
+	}
+
+	// When the mantissa already has the sign bit set, the number is too large
+	// to fit into the available 23-bits, so divide the number by 256 and
+	// increment the exponent accordingly.
+	if mantissa&0x00800000 != 0 {
+		mantissa >>= 8
+		exponent++
+	}
+
+	// Pack the exponent and mantissa into an unsigned 32-bit int and return it.
+	//
+	// Note that the sign bit is always unset here because uint256s can never be
```

Comment:
This is leftover from before the flag to force it to be negative was added.  It needs to be updated to reflect reality.

Reply:
Updated.